zig/private/providers/zig_module_info.bzl

```diff
@@ -36,111 +39,46 @@ def zig_module_info(*, name, canonical_name, main, srcs, extra_srcs, deps):
-    dep_spec = dep.name
-
-    if hasattr(dep, "canonical_name") and dep.canonical_name != dep.name:
-        dep_spec += "=" + dep.canonical_name
-
-    return dep_spec
-
-def _render_args(args):
-    rendered = []
+    return dep.name + "=" + dep.canonical_name
 
-    for dep in args.deps:
-        rendered.extend(["--dep", _render_dep(dep)])
+def _add_module_files(inputs, module):
+    deps = (module.main,) + module.srcs + module.extra_srcs
+    inputs.append(depset(direct = deps))
 
-    rendered.extend(["-M{name}={main}".format(
-        name = args.name,
-        main = args.main,
-    )])
+def _render_args(*, module, inputs, args):
+    args.add_all(module.deps, before_each = "--dep", map_each = _render_dep)
+    args.add(module.main, format = "-M{}=%s".format(module.canonical_name))
+    _add_module_files(inputs, module)
 
-    return rendered
-
-def zig_module_dependencies(*, zig_version, deps, extra_deps = [], args):
-    """Collect flags for the Zig main module to depend on other modules.
-
-    Args:
-      zig_version: string, The version of the Zig SDK.
-      deps: List of Target, Considers the targets that have a ZigModuleInfo provider.
-      extra_deps: List of ZigModuleInfo.
-      args: Args; mutable, Append the needed Zig compiler flags to this object.
-    """
-    _ = zig_version  # @unused
-    deps_args = []
-
-    modules = [
-        dep[ZigModuleInfo]
-        for dep in deps
-        if ZigModuleInfo in dep
-    ] + extra_deps
-
-    for module in modules:
-        deps_args.append(_render_dep(module))
-
-    args.add_all(deps_args, before_each = "--dep")
-
-def zig_module_specifications(*, zig_version, deps, extra_deps = [], inputs, args):
+def zig_module_specifications(*, root_module, inputs, args):
     """Collect inputs and flags to build Zig modules.
 
     Args:
-      zig_version: string, The version of the Zig SDK.
-      deps: List of Target, Considers the targets that have a ZigModuleInfo provider.
-      extra_deps: List of ZigModuleInfo.
-      inputs: List of depset of File; mutable, Append the needed inputs to this list.
-      args: Args; mutable, Append the needed Zig compiler flags to this object.
+        root_module: TODO(corentin)
+        inputs: List of depset of File; mutable, Append the needed inputs to this list.
+        args: Args; mutable, Append the needed Zig compiler flags to this object.
     """
-    _ = zig_version  # @unused
-    transitive_args = []
-
-    modules = [
-        dep[ZigModuleInfo]
-        for dep in deps
-        if ZigModuleInfo in dep
-    ] + extra_deps
-
-    for module in modules:
-        transitive_args.append(module.transitive_args)
-        inputs.append(module.transitive_inputs)
-
-    render_args = _render_args
-
-    args.add_all(depset(transitive = transitive_args), map_each = render_args)
+    _render_args(
+        module = root_module,
+        inputs = inputs,
+        args = args,
+    )
+    for dep in root_module.transitive_deps.to_list():
```

Comment:
`depset.to_list()` should be avoided, [see here](https://bazel.build/rules/performance#avoid-depset-to-list). The previous version of this code was careful to avoid that.

Reply:
I reverted to transitive arg structs, but in a way that still allows for a single arg rendering path.